frontend/renew_proxies.py

```diff
@@ -0,0 +1,151 @@
+        os.rename(self.output + '.new', self.output)
+
+    def timeleft(self):
+        """Returns the remaining lifetime of the proxy in seconds
+        """
+        stdout, _, _ = _run_command(['voms-proxy-info', '-file', self.output, '-timeleft'])
+        try:
+            return int(stdout)
+        except ValueError:
+            return 0
+
+class VO(object):
+    """Class for holding information related to VOMS attributes
+    """
+    def __init__(self, vo, command, cert=None, key=None):
+        self.name = vo
+        command = command.lstrip('/')
+        self.fqan = '/%s/%s' % (vo, command)
+        # intended argument for -voms option "vo:command" format, see voms-proxy-init man page
+        self.voms = '%s:/%s' % (vo, command)
+        self.cert = cert
+        self.key = key
+
+def _run_command(command):
+    """Runs the specified command, specified as a list. Returns stdout, stderr and return code
+    """
+    proc = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    stdout, stderr = proc.communicate()
+    return stdout, stderr, proc.returncode
+
+def voms_proxy_init(proxy, *args):
+    """Create a proxy using voms-proxy-init. Without any additional args, it generates a proxy without VOMS attributes.
+    Returns stdout, stderr, and return code of voms-proxy-init
+    """
+    cmd = ['voms-proxy-init', '--debug',
+           '-cert', proxy.cert,
+           '-key', proxy.key,
+           '-out', proxy.output + '.new',
+           '-valid', '%s:00' % proxy.lifetime] + list(args)
+    return _run_command(cmd)
+
+def voms_proxy_fake(proxy, vo_info, voms_uri):
+    """ Create a valid proxy without contacting a VOMS Admin server. VOMS attributes are created from user config.
+    Returns stdout, stderr, and return code of voms-proxy-fake
+    """
+    cmd = ['voms-proxy-fake', '--debug',
+           '-cert', proxy.cert,
+           '-key', proxy.key,
+           '-out', proxy.output + '.new',
+           '-hours', proxy.lifetime,
+           '-voms', vo_info.name,
+           '-hostcert', vo_info.cert,
+           '-hostkey', vo_info.key,
+           '-uri', voms_uri,
+           '-fqan', vo_info.fqan]
+    return _run_command(cmd)
+
+def main():
+    """Main entrypoint
+    """
+    config = ConfigParser.ConfigParser(DEFAULTS)
+    config.read(CONFIG)
+    proxies = config.sections()
+
+    # Verify config sections
+    if proxies.count('FRONTEND') != 1:
+        raise RuntimeError("ERROR: there must be only one [FRONTEND] section in %s" % CONFIG)
+    if len([x for x in proxies if x.startswith('PILOT')]) < 1:
+        raise RuntimeError("ERROR: there must be at least one [PILOT] section in %s" % CONFIG)
+
+    # Proxies need to be owned by the 'frontend' user
+    try:
+        fe_user = pwd.getpwnam('frontend')
+    except KeyError:
+        raise RuntimeError("ERROR: missing 'frontend' user")
+
+    # Load VOMS Admin server info for case-sensitive VO name and for faking the VOMS Admin server URI
```

Comment:
Hm - at some point, we'll need to think about how we get rid of the URI... I suppose that's not for today.